Add generate_stage_to_rank_mapping utility

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #146217
* __->__ #146193

We use `stage_index_to_group_rank` in the stage to determine what send/recv ops and in the schedule for IR generation. However, we don't need to expose this as an argument in our schedule class, so this stack of PRs is to remove it.

This PR creates a `stage_index_to_group_rank` utility function and removes the arg for the ZBVschedule. In a following PR I will add code to infer the `stage_index_to_group_rank` for the CSV schedule path and we will be able to remove this argument from our classes entirely.

Related comment from @wconstab https://github.com/pytorch/torchtitan/issues/774#issuecomment-2619793741

cc @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
Add generate_stage_to_rank_mapping utility

[ghstack-poisoned]

wconstab commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -1123,10 +1124,14 @@ def __init__(
-        self.stage_index_to_group_rank = stages[0].stage_index_to_group_rank
+        if stage_index_to_group_rank is None:
+            self.stage_index_to_group_rank = generate_stage_to_rank_mapping(
+                self.pp_group_size, self._num_stages
```
nit: might be worth being explicit that we're requesting "loop" here for readability and for not relying on a defaul

wconstab commented on test/distributed/pipelining/test_schedule.py:
```diff
@@ -928,6 +929,74 @@ def test_invalid_schedule_missing_action(self):
+                3: 0,
+            },
+        )
+        stage_to_rank = generate_stage_to_rank_mapping(4, 12, style="v")
```
Do V-style schedules allow double-looping?  I am not sure if even 4,16 is a valid config but 4,12 looks sus to me.  (But i might be wrong)

wconstab replied:
Ok. I guess I am ok with it if you document it

H-Huang committed:
Update on "Add generate_stage_to_rank_mapping utility"


We use `stage_index_to_group_rank` in the stage to determine what send/recv ops and in the schedule for IR generation. However, we don't need to expose this as an argument in our schedule class, so this stack of PRs is to remove it.

This PR creates a `stage_index_to_group_rank` utility function and removes the arg for the ZBVschedule. In a following PR I will add code to infer the `stage_index_to_group_rank` for the CSV schedule path and we will be able to remove this argument from our classes entirely.

Related comment from wconstab https://github.com/pytorch/torchtitan/issues/774#issuecomment-2619793741

cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]
